lib/decanter/parser/core.rb

```diff
@@ -8,38 +8,42 @@ def self.included(base)
 
       module ClassMethods
 
-        # Meant to be overriden and called by child parser
+        # Check if allowed, parse if not
         def parse(name, values, options={})
-
-          value_ary = values.is_a?(Array) ? values : [values]
-
-          # want to treat 'false' as an actual value
-          if value_ary.all? { |value| value.nil? || value == "" }
-            if options[:required]
-              raise ArgumentError.new("No value for required argument: #{name}")
-            else
-              return { name => nil }
-            end
-          end
-
-          if @allowed && value_ary.all? { |value| @allowed.any? { |allowed| value.is_a? allowed } }
-            return { name => values }
+          case
+          when allowed?(values)
+            { name => values }
+          else
+            _parse(name, values, options)
           end
-
-          unless @parser
-            raise ArgumentError.new("No parser for argument: #{name} with types: #{value_ary.map(&:class).join(', ')}")
-          end
-
-          _parse(name, values, options)
         end
 
+        # Define parser
         def parser(&block)
           @parser = block
         end
 
+        # Set allowed classes
         def allow(*args)
           @allowed = args
         end
+
+        # Set preparsers
+        def pre(*parsers)
+          @pre = parsers
+        end
+
+        # Get prepareer
+        def preparsers
+          @pre || []
+        end
+
+        # Check for allowed classes
+        def allowed?(values)
```

Comment:
Thoughts on using a more explicit guard like so? More of a pattern preference, no need to change the code for this PR.

```ruby
def allowed?(values)
  return false if @allowed.nil?
  Array.wrap(values).all? do |value|
    @allowed.any? { |allowed| value.is_a? allowed }
  end
end
```